cmd/thanos/query.go

```diff
@@ -210,21 +225,84 @@ func runQuery(
 	})
 	reg.MustRegister(duplicatedStores)
 
-	dialOpts, err := extgrpc.StoreClientGRPCOpts(logger, reg, tracer, secure, cert, key, caCert, serverName)
-	if err != nil {
-		return errors.Wrap(err, "building gRPC client")
+	var storesConfig []store.Config
+	if len(storeConfigYAML) > 0 {
+		var err error
+		storesConfig, err = store.LoadConfig(storeConfigYAML)
+		if err != nil {
+			return errors.Wrap(err, "loading store config")
+		}
+	} else {
+		for _, addr := range storeAddrs {
+			if addr == "" {
+				return errors.New("static store address cannot be empty")
+			}
+		}
+
+		endpointsConfig := store.EndpointsConfig{
+			StaticAddresses: storeAddrs,
+		}
+		if fileSDConfig != nil {
+			endpointsConfig.FileSDConfigs = []file.SDConfig{*fileSDConfig}
+		}
+		storeConfig := store.Config{
+			Name:            "default",
+			EndpointsConfig: endpointsConfig,
+		}
+		if secure {
+			storeConfig.TlsConfig = &store.TlsConfig{
+				Cert:       cert,
+				Key:        key,
+				CaCert:     caCert,
+				ServerName: serverName,
+			}
+		}
+
+		storesConfig = []store.Config{storeConfig}
 	}
 
-	fileSDCache := cache.New()
-	dnsProvider := dns.NewProvider(
-		logger,
-		extprom.WrapRegistererWithPrefix("thanos_querier_store_apis_", reg),
-		dns.ResolverType(dnsSDResolver),
+	var storeSets []*query.StoreSet
```

Comment:
`storeSets` is referenced in `allClients` but initialized later in the range over `storesConfig` which can be hard to reason about. Let me think if we can simplify this.